app/models/nic/base.rb

```diff
@@ -144,16 +150,30 @@ def host_managed?
       self.host && self.host.managed? && SETTINGS[:unattended]
     end
 
-    def require_ip_validation?
+    def require_ip_validation?(ip_field, other_ip_field, subnet)
+      raise ::Foreman::Exception.new("invalid ip field: %s", ip_field) unless ip_field == :ip || ip_field == :ip6
+      raise ::Foreman::Exception.new("invalid other ip field: %s", other_ip_field) unless other_ip_field == :ip || other_ip_field == :ip6
+
       # if it's not managed there's nowhere to specify an IP anyway
-      return false if !self.host.managed? || !self.managed? || !self.provision?
+      return false unless self.host.managed? && self.managed? && self.provision?
+
       # if the CR will provide an IP, then don't validate yet
-      return false if host.compute_provides?(:ip)
-      ip_for_dns     = (subnet.present? && subnet.dns_id.present?) || (domain.present? && domain.dns_id.present?)
-      ip_for_dhcp    = subnet.present? && subnet.dhcp_id.present?
-      ip_for_token   = Setting[:token_duration] == 0 && (host.pxe_build? || (host.image_build? && host.image.try(:user_data?)))
+      return false if host.compute_provides?(ip_field)
+
+      ip_for_dns   = (subnet.present? && subnet.dns_id.present?) || (domain.present? && domain.dns_id.present? && !self.send(other_ip_field).present?)
+      ip_for_dhcp  = subnet.present? && subnet.dhcp_id.present?
+      ip_for_token = Setting[:token_duration] == 0 && (host.pxe_build? || (host.image_build? && host.image.try(:user_data?))) && !self.send(other_ip_field).present?
+
       # Any of these conditions will require an IP, so chain with OR
-      ip_for_dns or ip_for_dhcp or ip_for_token
+      ip_for_dns || ip_for_dhcp || ip_for_token
+    end
+
+    def require_ip4_validation?
+      require_ip_validation?(:ip, :ip6, subnet)
```

Comment:
This and the whole `require_ip_validation?` method are shadowing the `subnet` from `belongs_to` https://github.com/theforeman/foreman/pull/3498/files#diff-30afe2d935f843af2132911906f3f1c7R61 - please find another name or extract the validation from this class to make it a bit less confusing